### PR TITLE
mon: add min_mon_release in monmap

### PR DESCRIPTION
**Pull request by liewegas**
- add min_mon_release to monmap
- integration with probe and election process
- automatically bump when we have a complete mon quorum that is a newer version

and finally,

- prevent mon from starting if monmap's min_mon_release is >2 releases old

### CODE REVIEW

**Comment by jdurgin** [comment]
would a mon store meta guard like you just added for the osd prevent potential rocksdb issues? it looks like read/write_meta work before opening the underlying db

**Comment by liewegas** [comment]
yeah! I didn't relaize MonitorDBStore had read_meta/write_meta.

File: src/mon/MonMap.cc [hunk 1]
```diff
@@ -220,6 +221,11 @@ void MonMap::decode(bufferlist::const_iterator& p)
   } else {
     decode(ranks, p);
   }
+  if (struct_v >= 6) {
```
Comment (tchaikov) [inline src/mon/MonMap.cc]: i think this should be `struct_v >= 7`.

File: src/tools/monmaptool.cc [hunk 2]
```diff
@@ -29,8 +29,9 @@ void usage()
        << "        [--add name 1.2.3.4:567] [--rm name]\n"
        << "        [--feature-list [plain|parseable]]\n"
        << "        [--feature-set <value> [--optional|--persistent]]\n"
-       << "        [--feature-unset <value> [--optional|--persistent]] "
-       << "<mapfilename>"
+       << "        [--feature-unset <value> [--optional|--persistent]]\n"
+       << "        [--set-min-mon-release <release-major-number>]\n"
+       << "        <mapfilename>"
```
Comment (tchaikov) [inline src/tools/monmaptool.cc]: might want to update `git grep '<mapfilename>'|grep cli`.

File: src/mon/MonmapMonitor.cc [hunk 3]
```diff
@@ -112,12 +112,13 @@ void MonmapMonitor::encode_pending(MonitorDBStore::TransactionRef t)
 class C_ApplyFeatures : public Context {
   MonmapMonitor *svc;
   mon_feature_t features;
-  public:
-  C_ApplyFeatures(MonmapMonitor *s, const mon_feature_t& f) :
-    svc(s), features(f) { }
+  int min_mon_release;
```
Comment (tchaikov) [inline src/mon/MonmapMonitor.cc]: nit, could make it `const`.

File: src/mon/MonmapMonitor.cc [hunk 4]
```diff
@@ -146,9 +148,9 @@ void MonmapMonitor::apply_mon_features(const mon_feature_t& features)
     (pending_map.persistent_features ^
      (features & ceph::features::mon::get_persistent()));
 
-  if (new_features.empty()) {
-    dout(10) << __func__ << " features match current pending: "
-             << features << dendl;
+  if (new_features.empty() && pending_map.min_mon_release == min_mon_release) {
```
Comment (tchaikov) [inline src/mon/MonmapMonitor.cc]: i'd suggest move the check of `pending_map.min_mon_release == min_mon_release` down before setting `pending_map.min_mon_release`. and 1) do not set `new_features` if it is empty,  2) do not update `pending_map.persistent_features` if it is empty. or better off guarding all of them with `if (new_features.empty()`. and set a flag `changed` to true if either `features` or `min_mon_release` is updated in the `pending_map`

File: src/ceph_mon.cc [hunk 5]
```diff
@@ -558,6 +558,26 @@ int main(int argc, const char **argv)
...
+    if (r >= 0 && val.size()) {
+      int min = atoi(val.c_str());
+      if (min &&
+	  min + 2 < (int)ceph_release()) {
+	derr << "recorded min_mon_release is " << min
+	     << " (" << ceph_release_name(min)
+	     << ") which is >2 releases older than current "
+	     << ceph_release() << " (" << ceph_release_name(ceph_release())
+	     << ")" << dendl;
+	derr << "you should first upgrade to an older release of ceph" << dendl;
```
[the remark this answers is not in the document]
Comment (liewegas) [reply src/ceph_mon.cc]: in this case, older.. they are upgrading >2 versions and we're too new.

File: src/mon/MonmapMonitor.cc (hunk 4 above)
Comment (liewegas) [reply src/mon/MonmapMonitor.cc]: cleaned up this function significantly.. take a look

File: src/ceph_mon.cc (hunk 5 above)
[the remark this answers is not in the document]
Comment (tchaikov) [reply src/ceph_mon.cc]: could list the 2 versions that are supported? namely, `ceph_release_name(min + 1)`, `ceph_release_name(min + 2)` ?

**Comment by tchaikov** [comment]
retest this please.

**Comment by liewegas** [comment]
retest this please

**Comment by tchaikov** [comment]
```
/home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos7/DIST/centos7/MACHINE_SIZE/huge/release/14.0.1-3673-ge7b0fd8/rpm/el7/BUILD/ceph-14.0.1-3673-ge7b0fd8/src/test/mon/test-mon-msg.cc:293:56: error: no matching function for call to 'MMonProbe::MMonProbe(uuid_d&, MMonProbe::<unnamed enum>, const char [2], bool)'
                         MMonProbe::OP_PROBE, "b", false);
                                                        ^
```

might want to update `src/test/mon/test-mon-msg.cc`. see https://jenkins.ceph.com/job/ceph-dev-new-build/ARCH=x86_64,AVAILABLE_ARCH=x86_64,AVAILABLE_DIST=centos7,DIST=centos7,MACHINE_SIZE=huge/17734//consoleFull

**Pull request merged**